plugin/trino-open-policy-agent/src/main/java/io/trino/plugin/openpolicyagent/OpaAuthorizer.java

```diff
@@ -0,0 +1,827 @@
+        if (!queryOpaWithSimpleResource(context, "InsertIntoTable", resource)) {
+            SystemAccessControl.super.checkCanInsertIntoTable(context, table);
+        }
+    }
+
+    @Override
+    public void checkCanDeleteFromTable(SystemSecurityContext context, CatalogSchemaTableName table)
+    {
+        OpaQueryInputResource resource = new OpaQueryInputResource.Builder().table(new OpaQueryInputResource.Table(table)).build();
+        if (!queryOpaWithSimpleResource(context, "DeleteFromTable", resource)) {
+            SystemAccessControl.super.checkCanDeleteFromTable(context, table);
+        }
+    }
+
+    @Override
+    public void checkCanTruncateTable(SystemSecurityContext context, CatalogSchemaTableName table)
+    {
+        OpaQueryInputResource resource = new OpaQueryInputResource.Builder().table(new OpaQueryInputResource.Table(table)).build();
+        if (!queryOpaWithSimpleResource(context, "TruncateTable", resource)) {
+            SystemAccessControl.super.checkCanTruncateTable(context, table);
+        }
+    }
+
+    @Override
+    public void checkCanUpdateTableColumns(SystemSecurityContext securityContext, CatalogSchemaTableName table, Set<String> updatedColumnNames)
+    {
+        OpaQueryInputResource resource = new OpaQueryInputResource.Builder().table(new OpaQueryInputResource.Table(table, updatedColumnNames)).build();
+
+        if (!queryOpaWithSimpleResource(securityContext, "UpdateTableColumns", resource)) {
+            SystemAccessControl.super.checkCanUpdateTableColumns(securityContext, table, updatedColumnNames);
+        }
+    }
+
+    @Override
+    public void checkCanCreateView(SystemSecurityContext context, CatalogSchemaTableName view)
+    {
+        OpaQueryInputResource resource = new OpaQueryInputResource.Builder().view(new OpaQueryInputResource.Table(view)).build();
+        if (!queryOpaWithSimpleResource(context, "CreateView", resource)) {
+            SystemAccessControl.super.checkCanCreateView(context, view);
+        }
+    }
+
+    @Override
+    public void checkCanRenameView(SystemSecurityContext context, CatalogSchemaTableName view, CatalogSchemaTableName newView)
+    {
+        OpaQueryInputResource resource = new OpaQueryInputResource.Builder().view(new OpaQueryInputResource.Table(view)).build();
+        OpaQueryInputResource targetResource = new OpaQueryInputResource.Builder().view(new OpaQueryInputResource.Table(newView)).build();
+        OpaQueryInputAction action = new OpaQueryInputAction.Builder().operation("RenameView").resource(resource).targetResource(targetResource).build();
+        OpaQueryInput input = new OpaQueryInput(context, action);
+
+        if (!queryOpa(input)) {
+            SystemAccessControl.super.checkCanRenameView(context, view, newView);
+        }
+    }
+
+    @Override
+    public void checkCanSetViewAuthorization(SystemSecurityContext context, CatalogSchemaTableName view, TrinoPrincipal principal)
```

Comment:
Setting view authorization is a thorny subject. It's surprisingly tricky to do in a way which avoids potential data leaks. The recently merged trinodb/trino#16691 is a result of almost two years of figuring out what to do with this :) Here we have something similar to what we did for file-based access control, so probably no action is required, but that's something to definitely keep in mind.

Reply:
Well, it depends, really, on how much flexibility we want to give. In the file-based access control in Trino we decided to add a new, specific privilege for changing authorization. In others it may require full admin access. We also discussed rules like "setting authorization to USER would require a privilege to impersonate that user, while setting it to ROLE would require a role grant with grant option" and enforcing them in the Trino engine; we eventually decided against the latter and to let each implementation choose its own way, instead. So I guess the way it is now gives maximum flexibility, but minimum guidance in avoiding pitfalls :) I'm just saying that this is a place where you have to be very careful.

Reply:
Got it. Should these nuances perhaps be documented on the javadoc of the SPI? Just to ensure future implementers are aware

Reply:
I guess it should. cc: @kokosing